docs/proposals/20200506-conditions.md

```diff
@@ -0,0 +1,550 @@
+  the core types, is highly correlated to the possibility to collect and summarize meaningful information
+  from the underlying hierarchy of objects.
+- Cluster API defines a set of core objects and extension points (e.g. bootstrap, control plane and infrastructure).
+  Extension points are defined by a set of contracts, which each provider adheres to. To provide a great user
+  experience, core and provider contracts should be extended to allow for better coordination and visibility
+  based on conditions.
+
+
+This proposal consists of three parts:
+
+- Data model changes.
+- Constraints/design principles for implementing conditions in Cluster API-
+- Use case driven examples of how conditions could improve observability on Cluster API.
+
+#### Data Model Changes
+
+In this section we outline the proposed API additions.
+
+All the changes can be implemented in the v0.3.x timeframe (no breaking changes, only additions).
+
+Following types should be defined in Cluster API / v1alpha3 api.
+
+```golang
+// ConditionType is a valid value for Condition.Type.
+type ConditionType string
+
+// ConditionSeverity expresses the severity of a Condition Type failing.
+type ConditionSeverity string
+
+const (
+  // ConditionSeverityError specifies that a failure of a condition type
+  // should be viewed as an error.
+  ConditionSeverityError ConditionSeverity = "Error"
+
+  // ConditionSeverityWarning specifies that a failure of a condition type
+  // should be viewed as a warning, but that things could still work.
+  ConditionSeverityWarning ConditionSeverity = "Warning"
+
+  // ConditionSeverityInfo specifies that a failure of a condition type
+  // should be viewed as purely informational, and that things could still work.
+  ConditionSeverityInfo ConditionSeverity = "Info"
+
+  // ConditionSeverityNone should apply only if the condition is in state "True".
+  // As ConditionSeverityNone is the default for conditions we use the empty string (coupled with omitempty)
+  ConditionSeverityNone ConditionSeverity = ""
+)
+
+// Condition defines an extension to status (i.e. an observation) of a Cluster API resource.
+type Condition struct {
+   // Type of condition.
+   // +required
+   Type ConditionType `json:"type" description:"type of status condition"`
+
+   // Status of the condition, one of True, False, Unknown.
+   // +required
+   Status corev1.ConditionStatus `json:"status"`
+
+   // Severity with which to treat failures of this type of condition.
+   // When this is not specified, it defaults to Error.
+   // +optional
+   Severity ConditionSeverity `json:"severity,omitempty"`
+
+   // LastTransitionTime is the last time the condition transitioned from one status to another.
+   // +required
+   LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+   // The reason for the condition's last transition.
+   // Reasons should be one-word CamelCase.
```

Comment:
I think this just should say "Reasons should be CamelCase". As written it sounds like they are supposed to be one word, which isn't true given the examples (`MachineNotHealthy` is three words). I think the fact that it's camelcase implies it's joined together without spaces.